pkg/collector/corechecks/cluster/orchestrator/discovery/collector_discovery.go

```diff
@@ -27,35 +28,107 @@ const (
 	defaultAPIServerTimeout = 20 * time.Second
 )
 
+type discoveryCache struct {
+	groups              []*v1.APIGroup
+	resources           []*v1.APIResourceList
+	filled              bool
+	collectorForVersion map[collectorVersion]struct{}
+}
+
+type collectorVersion struct {
+	version string
+	name    string
+}
+
 // APIServerDiscoveryProvider is a discovery provider that uses the Kubernetes
 // API Server as its data source.
 type APIServerDiscoveryProvider struct {
 	result []collectors.Collector
 	seen   map[string]struct{}
+	cache  discoveryCache
 }
 
 // NewAPIServerDiscoveryProvider returns a new instance of the APIServer
 // discovery provider.
 func NewAPIServerDiscoveryProvider() *APIServerDiscoveryProvider {
 	return &APIServerDiscoveryProvider{
-		seen: make(map[string]struct{}),
+		seen:  make(map[string]struct{}),
+		cache: discoveryCache{collectorForVersion: map[collectorVersion]struct{}{}}, // TODO: maybe dedicated file and make the discovery shared?
```

Comment:
I wasn't 100% sure whether this is the correct place to re-use this struct. Otherwise I had in mind to cut this out and make a dedicated struct and only re-use `getServerGroupsAndResources` + making that method to a function and public